src/test/java/com/greenfox/lvlup/service/BadgeServiceTest.java

```diff
@@ -0,0 +1,109 @@
+import com.greenfox.lvlup.model.dto.library.BadgeDTO;
+import com.greenfox.lvlup.model.entity.Badge;
+import com.greenfox.lvlup.model.entity.User;
+import com.greenfox.lvlup.repositrory.BadgeRepository;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.mockito.junit.MockitoJUnitRunner;
+import org.modelmapper.ModelMapper;
+
+import static org.junit.Assert.*;
+import static org.mockito.ArgumentMatchers.*;
+import static org.mockito.Mockito.*;
+
+
+@RunWith(MockitoJUnitRunner.class)
+public class BadgeServiceTest {
+    @InjectMocks
+    private BadgeService badgeService;
+
+    @Mock
+    private BadgeRepository badgeRepositoryMock;
+
+    @Mock
+    private ModelMapper modelMapperMock;
+
+    @Mock
+    private UserService userServiceMock;
+
+    User badgeCreator = new User("Test Elek");
+    String token = "testtoken";
+    String badToken = "notvalidtesttoken";
+    Badge validBadge = new Badge("2.3", "Test badge", "general");
+    Badge validBadge2 = new Badge("1.4", "English speaker", "general");
+    Badge validBadge3 = new Badge("1.5", "Feedback giver", "mentor");
+    BadgeDTO testDTO = new BadgeDTO("2.3", "Test badge", "general");
+
+
+    @Before
+    public void setUp() throws Exception {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void whenConvertBadgeDTOToBadge_thenReturnBadgeEntity() {
+        when(this.modelMapperMock.map(testDTO, Badge.class)).thenReturn(validBadge);
+        Badge result = badgeService.convertBadgeDTOToBadge(testDTO);
+
+        assertEquals(result.getName(), validBadge.getName());
+        assertEquals(result.getVersion(), validBadge.getVersion());
+        assertEquals(result.getTag(), validBadge.getTag());
+        assertEquals(result.getDateOfCreation(), validBadge.getDateOfCreation());
+        assertEquals(result.getUser(), validBadge.getUser());
+    }
+
+    @Test
+    public void whenFindBadgeByNameAndVersionExistingBadgeVersion_thenReturnExistingBadgeVersion() {
+        when(badgeRepositoryMock.findBadgeByNameAndVersion(anyString(), anyString())).thenReturn(validBadge);
+        Badge testResult = badgeService.findBadgeByNameAndVersion(validBadge);
+
+        assertNotNull(testResult);
+        assertEquals(validBadge.getVersion(), testResult.getVersion());
+        assertEquals(validBadge.getName(), testResult.getName());
+    }
+
+    @Test
+    public void whenFindBadgeByNameAndVersionWithNotExistingBadgeVersion_thenReturnNull() {
+        when(badgeRepositoryMock.findBadgeByNameAndVersion(anyString(), anyString())).thenReturn(null);
+        Badge emptyBadge = badgeService.findBadgeByNameAndVersion(validBadge);
+
+        assertNull(emptyBadge);
+    }
+
+    @Test
+    public void saveBadgeIntoDatabase() throws Exception {
+        when(badgeService.findBadgeByNameAndVersion(validBadge2)).thenReturn(null);
```

Comment:
In this test it looks like you are mocking something, then testing something else?

Reply:
Saving the badge into the Db wont return anything, since it's a void method, and it's hard to test it like this. It's better to return the created object, if it's in the db and then you can assert the result

Reply:
As discussed ealier I just check whether to badgeversion already exists and if not then  I will save it (with user) into database.